### PR TITLE
spawn-fcgi should be the prefered way compare to fcgiwrap

### PR DESCRIPTION
**Pull request by Gustry**
Goal:

A few days ago, I quickly followed the documentation with NGINX and I installed `fcgiwrap`. I didn't see the previous note made by @elemoine. The performance is catastrophic :) according to me. It's indeed very slow. So I think we should move `spawn-fcgi` before `fcgiwrap` in the documentation.

fcgiwrap is easy to setup, but very slow.

I also tried to add a systemd file for starting qgis server.

@elemoine I'm clearly not an expert in this area, a review is welcome.

Ticket(s): #

- [x] Backport to LTR documentation is required

### CODE REVIEW

File: source/docs/user_manual/working_with_ogc/server/getting_started.rst [hunk 1]
```diff
@@ -207,6 +172,85 @@ QGIS Server is now available at http://localhost/qgisserver.
...
+
+As mentioned above, we can setup `multiwatch` to increase the number of process to 2 for instance:
+
+.. code-block:: bash
+
+   sudo apt install multiwatch
+
+Either in your bash line or in your init file, launch your QGIS Server by adding this line before the path to the :file:`qgis_mapserv.fcgi`:
+
+..code-block:: bash
```
Comment (DelazJ) [inline source/docs/user_manual/working_with_ogc/server/getting_started.rst]:
```suggestion
.. code-block:: bash
```

File: source/docs/user_manual/working_with_ogc/server/getting_started.rst [hunk 2]
```diff
@@ -207,6 +172,85 @@ QGIS Server is now available at http://localhost/qgisserver.
...
+.. code-block:: bash
+
+   sudo systemctl enable /etc/systemd/system/qgis-server.service
+   sudo systemctl start qgis-server
+   sudo systemctl status qgis-server
+
+
+QGIS Server should be available and you can watch the live logs with:
+
+.. code-block::bash
```
Comment (DelazJ) [inline source/docs/user_manual/working_with_ogc/server/getting_started.rst]:
```suggestion
.. code-block:: bash
```

File: source/docs/user_manual/working_with_ogc/server/getting_started.rst [hunk 3]
```diff
@@ -207,6 +172,85 @@ QGIS Server is now available at http://localhost/qgisserver.
...
+
+   sudo journalctl -f -u qgis-server
+
+As mentioned above, we can setup `multiwatch` to increase the number of process to 2 for instance:
```
Comment (DelazJ) [inline source/docs/user_manual/working_with_ogc/server/getting_started.rst]:
```suggestion
As mentioned above, we can setup ``multiwatch`` to increase
the number of process to 2 for instance:
```
If you want it with a particular display

File: source/docs/user_manual/working_with_ogc/server/getting_started.rst [hunk 4]
```diff
@@ -207,6 +172,85 @@ QGIS Server is now available at http://localhost/qgisserver.
...
+
+.. code-block:: bash
+
+   sudo apt install multiwatch
+
+Either in your bash line or in your init file, launch your QGIS Server by adding this line before the path to the :file:`qgis_mapserv.fcgi`:
```
Comment (DelazJ) [inline source/docs/user_manual/working_with_ogc/server/getting_started.rst]:
```suggestion
Either in your bash line or in your init file, launch your QGIS Server
by adding this line before the path to the :file:`qgis_mapserv.fcgi`:
```
A bit too long

File: source/docs/user_manual/working_with_ogc/server/getting_started.rst [hunk 5]
```diff
@@ -207,6 +172,85 @@ QGIS Server is now available at http://localhost/qgisserver.
     `multiwatch <https://redmine.lighttpd.net/projects/multiwatch/wiki>`_ tool, which is also
     packaged in Debian.
 
+Of course, we can add an init script (like a ``qgis-server.service`` file
```
Comment (DelazJ) [inline source/docs/user_manual/working_with_ogc/server/getting_started.rst]: Why moving from "you" to "we"?
```suggestion
Of course, you can add an init script (like a ``qgis-server.service`` file
```

**Comment by DelazJ** [comment]
@elemoine may I?

**Comment by Gustry** [comment]
Maybe @pblottiere can give a review?

File: source/docs/user_manual/working_with_ogc/server/getting_started.rst [hunk 6]
```diff
@@ -207,6 +172,87 @@ QGIS Server is now available at http://localhost/qgisserver.
...
+
+
+QGIS Server should be available and you can watch the live logs with:
+
+.. code-block:: bash
+
+   sudo journalctl -f -u qgis-server
+
+As mentioned above, we can setup ``multiwatch`` to increase
+the number of process to 2 for instance:
```
Comment (pblottiere) [inline source/docs/user_manual/working_with_ogc/server/getting_started.rst]:
```suggestion
the number of processes to 2 for instance:
```

File: source/docs/user_manual/working_with_ogc/server/getting_started.rst [hunk 7]
```diff
@@ -207,6 +172,87 @@ QGIS Server is now available at http://localhost/qgisserver.
...
+
+.. code-block:: bash
+
+   sudo apt install multiwatch
+
+Either in your bash line or in your init file, launch your QGIS Server
+by adding this line before the path to the :file:`qgis_mapserv.fcgi`:
```
Comment (pblottiere) [inline source/docs/user_manual/working_with_ogc/server/getting_started.rst]:
```suggestion
by adding this line before the path to :file:`qgis_mapserv.fcgi`:
```

File: source/docs/user_manual/working_with_ogc/server/getting_started.rst [hunk 8]
```diff
@@ -207,6 +172,87 @@ QGIS Server is now available at http://localhost/qgisserver.
...
+   :linenos:
+
+     location /qgisserver {
+         gzip           off;
+         include        fastcgi_params;
+         fastcgi_pass   unix:/var/run/fcgiwrap.socket;
+         fastcgi_param  SCRIPT_FILENAME /usr/lib/cgi-bin/qgis_mapserv.fcgi;
+     }
+
+Finally, restart NGINX and fcgiwrap to take into account the new configuration:
```
Comment (pblottiere) [inline source/docs/user_manual/working_with_ogc/server/getting_started.rst]:
```suggestion
Finally, restart NGINX and **fcgiwrap** to take into account the new configuration:
```

Comment (pblottiere) [reply source/docs/user_manual/working_with_ogc/server/getting_started.rst]: (to be consistent with previous chapters)

File: source/docs/user_manual/working_with_ogc/server/getting_started.rst [hunk 9]
```diff
@@ -207,6 +172,87 @@ QGIS Server is now available at http://localhost/qgisserver.
     `multiwatch <https://redmine.lighttpd.net/projects/multiwatch/wiki>`_ tool, which is also
     packaged in Debian.
 
+Of course, you can add an init script (like a ``qgis-server.service`` file
+with Systemd) to start QGIS Server at boot time or whenever you want. To use a init script with systemd,
```
Comment (pblottiere) [inline source/docs/user_manual/working_with_ogc/server/getting_started.rst]:
```suggestion
with Systemd) to start QGIS Server at boot time or whenever you want. To use an init script with systemd,
```

File: source/docs/user_manual/working_with_ogc/server/getting_started.rst (hunk 7 above)
Comment (pblottiere) [reply source/docs/user_manual/working_with_ogc/server/getting_started.rst]: I think we should find a better wording for this sentence because it seems a little unclear.

File: source/docs/user_manual/working_with_ogc/server/getting_started.rst [hunk 10]
```diff
@@ -207,6 +172,87 @@ QGIS Server is now available at http://localhost/qgisserver.
...
+
+.. code-block:: bash
+
+   /usr/bin/multiwatch -f 2
+
+fcgiwrap
+^^^^^^^^
+
+Using `fcgiwrap <https://www.nginx.com/resources/wiki/start/topics/examples/fcgiwrap/>`_
+is much easier to setup than `spawn-fcgi` but it's much slower.
```
Comment (pblottiere) [inline source/docs/user_manual/working_with_ogc/server/getting_started.rst]:
```suggestion
is much easier to setup than **spawn-fcgi** but it's much slower.
```

Comment (pblottiere) [reply source/docs/user_manual/working_with_ogc/server/getting_started.rst]: for consistency

File: source/docs/user_manual/working_with_ogc/server/getting_started.rst [hunk 11]
```diff
@@ -207,6 +172,87 @@ QGIS Server is now available at http://localhost/qgisserver.
...
+   sudo systemctl status qgis-server
+
+
+QGIS Server should be available and you can watch the live logs with:
+
+.. code-block:: bash
+
+   sudo journalctl -f -u qgis-server
+
+As mentioned above, we can setup ``multiwatch`` to increase
```
Comment (pblottiere) [inline source/docs/user_manual/working_with_ogc/server/getting_started.rst]: maybe `**multiwatch**` to be consistent with `fcgiwrap`? Or is it the other way around? 

@DelazJ what should be preferred between **mytool** and ``mytool``?

**Comment by pblottiere** [comment]
> Maybe @pblottiere can give a review?

I just left a few comments.

And thanks @Gustry, I think it's a good move :+1:

File: source/docs/user_manual/working_with_ogc/server/getting_started.rst (hunk 9 above)
Comment (elemoine) [reply source/docs/user_manual/working_with_ogc/server/getting_started.rst]: s/Systemd/systemd to use the same spelling everywhere. And the official is "systemd" I guess.

File: source/docs/user_manual/working_with_ogc/server/getting_started.rst [hunk 12]
```diff
@@ -207,6 +172,87 @@ QGIS Server is now available at http://localhost/qgisserver.
...
+   [Service]
+   # Add some environment variable if needed
+   Environment=QGIS_SERVER_LOG_LEVEL=0
+
+   StandardOutput=null
+   StandardError=journal
+   ExecStart=spawn-fcgi -s /var/run/qgisserver.socket -U www-data -G www-data -n /usr/lib/cgi-bin/qgis_mapserv.fcgi
+
+   [Install]
+   WantedBy=multi-user.target
```
Comment (elemoine) [inline source/docs/user_manual/working_with_ogc/server/getting_started.rst]: I personally wouldn't provide a systemd service file in the QGIS documentation. I think this is going too far, with materials that are outside the scope of QGIS.

File: source/docs/user_manual/working_with_ogc/server/getting_started.rst (hunk 11 above)
Comment (elemoine) [reply source/docs/user_manual/working_with_ogc/server/getting_started.rst]: I actually would not mention multiwatch. I've had bad experience with it. And one could use a systemd template file for multi-processing. But again I think this is going too far the QGIS Server documentation.

**Comment by elemoine** [comment]
I just left a few comments. These are just my suggestions, take them the way you want :)

File: source/docs/user_manual/working_with_ogc/server/getting_started.rst (hunk 9 above)
Comment (pblottiere) [reply source/docs/user_manual/working_with_ogc/server/getting_started.rst]: Yes you're right, I updated the suggestion.

File: source/docs/user_manual/working_with_ogc/server/getting_started.rst (hunk 11 above)
Comment (pblottiere) [reply source/docs/user_manual/working_with_ogc/server/getting_started.rst]: >  I've had bad experience with it.

+1 to remove multiwatch from the documentation then

File: source/docs/user_manual/working_with_ogc/server/getting_started.rst (hunk 12 above)
Comment (pblottiere) [reply source/docs/user_manual/working_with_ogc/server/getting_started.rst]: @elemoine Mmmm, I see you're point. In this case, we could add the systemd service file in the training manual instead.

**Comment by pblottiere** [comment]
> fcgiwrap is easy to setup, but very slow.

Otherwise, after some reflection, I'm wondering if `fcgiwarp` should really be mentioned in the documentation. I think it's probably counterproductive to mention a solution which is considered as slow.

@elemoine @Gustry what do you think?

**Comment by elemoine** [comment]
> @elemoine @Gustry what do you think?

I'd keep it, as it's so straightforward to set up on Linux.

File: source/docs/user_manual/working_with_ogc/server/getting_started.rst (hunk 12 above)
Comment (elemoine) [reply source/docs/user_manual/working_with_ogc/server/getting_started.rst]: Yes, there are so many ways to set up QGIS Server, and FastCGI applications in general. The QGIS documentation can't cover it all. The current QGIS Server doc covers Apache + mod_fcgid, and NGINX + fcgiwrap and spawn-fcgi. That sounds sufficient to me.

**Comment by pblottiere** [comment]
> I'd keep it, as it's so straightforward to set up on Linux.

I'm still not convinced to mention a solution which should not be used in production :).

**Comment by Gustry** [comment]
I have just updated the PR.

I have no idea if we should keep fcgiwrap.

About the service file, I have removed it. I know the QGIS documentation can't (and mustn't) cover the doc of systemd, multiwatch, spawn-fcgi etc But it's not convenient to go from doc to doc and have always something more to read. That's why I included at the beginning a small service file template that we could easily copy/paste. I don't have a proper solution.

**Comment by DelazJ** [comment]
Hi all,
Having a look to the current diff, it's more about moving down the fcgiwrap description and raising from note to warning its conditions of usage. This looks like a good compromise for the issue about mentioning or not that system (I have no idea what you are talking about so no opinion myself).
Any objection to merging?

**Comment by elemoine** [comment]
Looks good to me.
